File: RoboLabWP/RoboLabWP/Model/Game/MZFlow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using RoboLabWP.Model.Basic_Types;



namespace RoboLabWP.Model.Game
{
    public class MZFlow
    {
        MZDirection _direction;
        MZPosition _position;
        int _length;
        double _speed;
        bool _isInUse;
        int _frameUpdated;
        double _shift;

        public MZFlow(MZPosition pos, MZDirection dir, int length, double speed)
        {
            _position = pos;
            _direction = dir;
            _length = length;
            _speed = speed;
            _isInUse = false;
        }

        public MZDirection direction()
        {
            return _direction;
        }

        public MZPosition position()
        {
            return _position;
        }

        public int length()
        {
            return _length;
        }

        public double speed()
        {
            return _speed;
        }

        public double shift()
        {
            return _shift;
        }

        public bool isInFlow(MZPosition position)
        {
            switch (_direction)
            {
                case MZDirection.UP:
                    if(_position.x() != position.x())
                        return false;
                    else
                    {
                        if(_position.y() <= position.y() && position.y() < _position.y() + _length - 1)
                            return true;
                        else
                            return false;
                    }
                case MZDirection.RIGHT:
                    if(_position.y() != position.y())
                        return false;
                    else
                    {
                        if(_position.x() <= position.x() && position.x() < _position.x() + _length - 1)
                            return true;
                        else
                            return false;
                    }
                case MZDirection.DOWN:
                    if(_position.x() != position.x())
                        return false;
                    else
                    {
                        if(_position.y() >= position.y() && position.y() > _position.y() - _length + 1)
                            return true;
                        else
                            return false;
                    }
                case MZDirection.LEFT:
                    if(_position.y() != position.y())
                        return false;
                    else
                    {
                        if(_position.x() >= position.x() && position.x() > _position.x() - _length + 1)
                            return true;
                        else
                            return false;
                    }
                default:
                    return false;
            }
        }

        public MZMotion moveInFlowWithPositionandFrame(MZPosition position, int frame)
        {
            if(!isInFlow(position))
            {
                _isInUse = false;
                _shift = 0;
                return null;
            }
            if(!_isInUse)
            {
                _isInUse = false;
                _frameUpdated = frame;
                position.move(_direction);
                 _shift = 0;
                return new  MZMotion(_speed, _direction);
            }
            else
            {
                if(frame - _frameUpdated >= 1.0 / _speed)
                {
                    _frameUpdated = frame;
                    position.move(_direction);
                    return new MZMotion(_speed, _direction);
                }
                _shift = (double)(frame - _frameUpdated) * _speed;
            }
            return null;
        }
    }
}

File: RoboLabWP/RoboLabWPComp/Model/Game/Game Objects/MZFlow.cpp
#include "pch.h"
#include "MZFlow.h"


MZFlow::MZFlow(MZPosition * pos, MZDirection dir, int length, double speed)
{
	_position = new MZPosition(*pos);
    _direction = dir;
    _length = length;
    _speed = speed;
	_isInUse = false;
}

MZFlow::~MZFlow(void)
{
	if(_position != NULL)
	{
		delete(_position);
		_position = NULL;
	}
}

MZPosition* MZFlow::position()
{
	return _position;
}
MZDirection MZFlow::direction()
{
	return _direction;
}

int MZFlow::length()
{
	return _length;
}
double MZFlow::speed()
{
	return _speed;
}
MZMotion* MZFlow::moveInFlowWithPositionandFrame(MZPosition* position, int frame)
{
	if(!isInFlow(position))
    {
        _isInUse = false;
        _shift = 0;
		return NULL;
    }
    if(!_isInUse)
    {
		_isInUse = true;
        _frameUpdated = frame;
        position->move(_direction);
         _shift = 0;
		 return new MZMotion(_speed,_direction);
    }
    else
    {
        if(frame - _frameUpdated >= 1.0 / _speed)
        {
            _frameUpdated = frame;
            position->move(_direction);
            return new MZMotion(_speed,_direction);
        }
        _shift = (double)(frame - _frameUpdated) * _speed;
    }
	return NULL;
}
bool MZFlow::isInFlow(MZPosition *position)
{
	switch (_direction)
    {
        case UP:
			if(_position->x() != position->x())
                return false;
            else
            {
				if((_position->y() <= position->y()) && (position->y() < _position->y() + _length - 1))
					return true;
                else
					return false;
            }
            break;
        case RIGHT:
            if(_position->y() != position->y())
                return false;
            else
            {
                if((_position->x() <= position->x()) && (position->x() < _position->x() + _length - 1))
                    return true;
                else
                    return false;
            }
            break;
        case DOWN:
            if(_position->x() != position->x())
				return false;
            else
            {
                if((_position->y() >= position->y()) && (position->y() > _position->y() - _length + 1))
                    return true;
                else
                    return false;
            }
            break;
        case LEFT:
            if(_position->y() != position->y())
                return false;
            else
            {
                if((_position->x() >= position->x()) && (position->x() > _position->x() - _length + 1))
                    return true;
                else
                    return false;
            }
        break;
    }
	return false;
}
double MZFlow::shift()
{
	return _shift;
}
void MZFlow::toggle()
{
	int x = _position->x();
	int y = _position->y();
	if(_position != NULL)
	{
		delete(_position);
		_position = NULL;
	}
    switch (_direction)
    {
        case LEFT:
            //_position = [[MZPosition alloc] initWithX:[_position x] - _length + 1 andY:[_position y]];
			x = x - _length + 1;
            break;
        case RIGHT:
            //_position = [[MZPosition alloc] initWithX:[_position x] + _length  - 1 andY:[_position y]];
			x = x + _length - 1;
			break;
        case UP:
            //_position = [[MZPosition alloc] initWithX:[_position x] andY:[_position y] + _length - 1];
			y = y + _length - 1;
            break;
        case DOWN:
            //_position = [[MZPosition alloc] initWithX:[_position x] andY:[_position y] - _length + 1];
			y = y - _length + 1;
            break;
    }
	_position = new MZPosition(x,y);
	_direction = (MZDirection)((_direction + 2) %4);
}